### PR TITLE
Added custom messages from Confluence

### PR DESCRIPTION
**Pull request by colinli02**
https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1764623016/SOFT+Integration+Planning#Custom-Messages 

### Description

- There is no open issue.
- Refer to items here for "Colin" https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/edit-v2/1798504659

### Verification

- Built in ROS (sailbot_workspace) and passed.
image

### CODE REVIEW

**Review by patrick-5546 — DISMISSED** [review]
Please add autogenerated files/directories to a `.gitignore` file loll

**Review by patrick-5546 — COMMENTED** [review]
I think we should make this our working spec, to replace https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1764623016/SOFT+Integration+Planning#Interfaces. This way we don't have to update the confluence page during development when we want to make a change to a msg. To do so, we should comment descriptions for fields that are not obvious: conventions, range, etc..
- `speed` and `NUM_WIND_SENSORS` are examples of fields that are obvious
- `direction_degrees` and `voltage` are examples of fields that are not obvious

@hhenry01 this is why I was thinking it would be good to have `Speed` and `Heading` helper messages: we would only have to write the conventions in one place, rather than every time it is used (GPS, Ship, Wind).

File: msg/LatLon.msg
Comment (patrick-5546) [inline msg/LatLon.msg:1]: Let's put internal/helper msgs like `LatLon` in a subdirectory

Comment (colinli02) [reply msg/LatLon.msg:1]: Added helpermsg/ within msg e.g.:
```
msg/helpermsg/LatLon.msg
```

**Comment by colinli02** [comment]
Made changes today. See:
https://github.com/UBCSailbot/ros_msgs/tree/remotes/origin/user/colinli02/custom_messages

Verified by passing build.

**Comment by colinli02** [comment]
### In case anyone sees the dismissed review above:

I dismissed an old review message about adding a .gitignore that was based on an old commit that has been rebased out of log. Had some auto-generated files from an initial creation of tutorial_interfaces (now ros_msgs). These files don't appear anymore so it's all good.

**Review by hhenry01 — APPROVED** [review]

**Review by patrick-5546 — COMMENTED** [review]
On the confluence page, I see that the comments about having uppercase field names are unresolved:

- @hhenry01 can we change them?
- @colinli02 please update the confluence page according to Henry's response

Meeting update: get rid of constant fields, fix size of array

~Also Colin just before you merge in this PR, can you publish a new version of the confluence page with a version comment that contains a link to this PR or the PR number?~ Instead, let's stop using the confluence page once this PR is merged in. This repository will serve as the spec for our interfaces. Please remind me to archive the page once this PR is merged in.

File: msg/LPathStats.msg
Comment (patrick-5546) [inline msg/LPathStats.msg:1]: Let's delete this file for now, we can create it once we know what its fields are

File: srv/Example.srv
Comment (patrick-5546) [inline srv/Example.srv:1]: Let's delete the srv directory, it only contains an example file

File: msg/Ship.msg
Comment (patrick-5546) [inline msg/Ship.msg]: Shouldn't the speed and heading fields be `helpermsg` types?

File: msg/helpermsg/Vector3.msg
Comment (patrick-5546) [inline msg/helpermsg/Vector3.msg:1]: I'm thinking this repository should only contain msgs that are used across ROS packages (interfaces). If they are used only in one package, it should be defined in its `msg/` directory. Also maybe we should rename this repository `interfaces`. @colinli02 can you create a discussion topic in next software wide meeting to talk about this (and any questions you have about my review)?

Comment (patrick-5546) [reply msg/helpermsg/Vector3.msg:1]: @DFriend01

Comment (colinli02) [reply msg/helpermsg/Vector3.msg:1]: # To-do for me (based on today 2023-06-1 meeting)
- [x] Rename ros_msgs -> ros_interface
- [x] ~~Directly call helper message (try helpermsg/Speed)~~ See comments from last week, we get rid of /helpermsg

Comment (patrick-5546) [reply msg/helpermsg/Vector3.msg:1]: Can rename repository `custom_interfaces` instead of `ros_interface`. ~Also, do you want to rename the repository? It would involve renaming this GitHub repository and updating `new_project.repos` and the workspace file in sailbot workspace accordingly similar to what you did you add the `ros_msgs` repo.~ I renamed the repository to unblock myself on writing the local pathfinding skeleton code.

**Comment by patrick-5546** [comment]
Just discovered this repository: https://github.com/ros2/common_interfaces/tree/humble

It defines some commonly used messages, including
- sensors, including GPS
- geometry, including vectors
- std: int, float, etc.

It may be worth taking a look at what the ros library offers before create our own custom types. And even if we can't use them directly we can draw inspiration from them.

I commented the lines that are required to add the `geometry_msgs` dependency as an example.

File: msg/helpermsg/Heading.msg
Comment (patrick-5546) [inline msg/helpermsg/Heading.msg:1]: ~Just realized that subdirectories of `msg/` and `srv` aren't supported: https://github.com/ros2/rosidl/issues/213.~ Nvm they are, but the interface names don't reflect its path. I think we should match that in this repo: get rid of `helpermsg/` and move the helper messages to `msg/`.

This can be verified through the `ros interface show` command:
image

File: msg/Path.msg
Comment (patrick-5546) [inline msg/Path.msg:1]: Can you rename this to `DesiredHeading.msg`

File: msg/AIS.msg
Comment (patrick-5546) [inline msg/AIS.msg:1]: Rename to AISShips.msg

File: msg/Ship.msg
Comment (patrick-5546) [inline msg/Ship.msg:1]: Rename to AISShip.msg

File: msg/Wind.msg
Comment (patrick-5546) [inline msg/Wind.msg:1]: Rename to WindSensors.msg

**Review by patrick-5546 — COMMENTED** [review]
More renaming

File: msg/helpermsg/Heading.msg
Comment (patrick-5546) [reply msg/helpermsg/Heading.msg:1]: You can also try to find standard library msgs instead of defining our own helper messages: https://github.com/UBCSailbot/custom_interfaces/pull/2#issuecomment-1577866615

File: msg/Wind.msg
Comment (colinli02) [reply msg/Wind.msg:1]: Renamed

File: msg/Ship.msg
Comment (colinli02) [reply msg/Ship.msg:1]: Renamed

File: msg/AIS.msg
Comment (colinli02) [reply msg/AIS.msg:1]: Renamed

File: msg/Path.msg
Comment (colinli02) [reply msg/Path.msg:1]: re-named

**Comment by colinli02** [comment]
Future to-do: look at existing libraries https://github.com/UBCSailbot/custom_interfaces/pull/2#issuecomment-1577866615

**Comment by colinli02** [comment]
# EDIT: Resolved. I forgot confluence page != PR 
# See bottom of comment

Also super weird, I don't know what's causing this but I noticed this last time and this time, ```Ship.msg``` must exist or else ```custom_interfaces``` will not build. 

See error message:
```Executing task in folder sailbot_workspace: ./build.sh 

Starting >>> custom_interfaces
Starting >>> network_systems
Finished <<< network_systems [0.16s]                                                                  
--- stderr: custom_interfaces                            
In file included from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_c/custom_interfaces/msg/detail/ais_ships__functions.h:19,
                 from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_c/custom_interfaces/msg/detail/ais_ships__functions.c:4:
/workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_c/custom_interfaces/msg/detail/ais_ships__struct.h:22:10: fatal error: custom_interfaces/msg/detail/ship__struct.h: No such file or directory
   22 | #include "custom_interfaces/msg/detail/ship__struct.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/custom_interfaces__rosidl_generator_c.dir/build.make:231: CMakeFiles/custom_interfaces__rosidl_generator_c.dir/rosidl_generator_c/custom_interfaces/msg/detail/ais_ships__functions.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:213: CMakeFiles/custom_interfaces__rosidl_generator_c.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
In file included from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_typesupport_cpp/custom_interfaces/msg/ais_ships__type_support.cpp:7:
/workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_cpp/custom_interfaces/msg/detail/ais_ships__struct.hpp:20:10: fatal error: custom_interfaces/msg/detail/ship__struct.hpp: No such file or directory
   20 | #include "custom_interfaces/msg/detail/ship__struct.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_typesupport_fastrtps_cpp/custom_interfaces/msg/detail/ais_ships__rosidl_typesupport_fastrtps_cpp.hpp:11,
                 from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_typesupport_fastrtps_cpp/custom_interfaces/msg/detail/dds_fastrtps/ais_ships__type_support.cpp:4:
/workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_cpp/custom_interfaces/msg/detail/ais_ships__struct.hpp:20:10: fatal error: custom_interfaces/msg/detail/ship__struct.hpp: No such file or directory
   20 | #include "custom_interfaces/msg/detail/ship__struct.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/custom_interfaces__rosidl_typesupport_cpp.dir/build.make:128: CMakeFiles/custom_interfaces__rosidl_typesupport_cpp.dir/rosidl_typesupport_cpp/custom_interfaces/msg/ais_ships__type_support.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:395: CMakeFiles/custom_interfaces__rosidl_typesupport_cpp.dir/all] Error 2
gmake[2]: *** [CMakeFiles/custom_interfaces__rosidl_typesupport_fastrtps_cpp.dir/build.make:154: CMakeFiles/custom_interfaces__rosidl_typesupport_fastrtps_cpp.dir/rosidl_typesupport_fastrtps_cpp/custom_interfaces/msg/detail/dds_fastrtps/ais_ships__type_support.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
In file included from /workspaces/sailbot_workspace/build/custom_interfaces/rosidl_typesupport_introspection_cpp/custom_interfaces/msg/detail/ais_ships__type_support.cpp:12:
/workspaces/sailbot_workspace/build/custom_interfaces/rosidl_generator_cpp/custom_interfaces/msg/detail/ais_ships__struct.hpp:20:10: fatal error: custom_interfaces/msg/detail/ship__struct.hpp: No such file or directory
   20 | #include "custom_interfaces/msg/detail/ship__struct.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/custom_interfaces__rosidl_typesupport_introspection_cpp.dir/build.make:154: CMakeFiles/custom_interfaces__rosidl_typesupport_introspection_cpp.dir/rosidl_typesupport_introspection_cpp/custom_interfaces/msg/detail/ais_ships__type_support.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:369: CMakeFiles/custom_interfaces__rosidl_typesupport_introspection_cpp.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:343: CMakeFiles/custom_interfaces__rosidl_typesupport_fastrtps_cpp.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< custom_interfaces [1.36s, exited with code 2]

Summary: 1 package finished [1.62s]
  1 package failed: custom_interfaces
  1 package had stderr output: custom_interfaces
  ```
  
  Resolved by making an empty "Ship.msg". Why? Note this doesn't happen to any other .msg file.
  I made sure to remove Ship.msg from the cmake file, even cleared the build file, still this happens.
  
  Temporary solution was to add an empty Ship.msg so it builds for now.
  
  # Edit resolved
  - Forgot PR comments diverged from confluence documentation
  - Solution: ```Ship[] ships -> AISShip[] ships```

File: msg/helpermsg/Heading.msg
Comment (colinli02) [reply msg/helpermsg/Heading.msg:1]: I've moved from ```helpermsg/``` to ```msg/``` . For some reason that command doesnt work (ros2 interface show custom_interfaces/)
image

So i manually verified using ```ros2 interface show <msg_type>```
image

File: srv/Example.srv
Comment (colinli02) [reply srv/Example.srv:1]: /srv deleted

File: msg/LPathStats.msg
Comment (colinli02) [reply msg/LPathStats.msg:1]: Deleted

File: msg/Ship.msg
Comment (colinli02) [reply msg/Ship.msg]: Fixed all to be helper msg type e.g.
```
LatLon lat_lon
Speed speed
Heading heading
```

File: msg/helpermsg/Heading.msg
Comment (patrick-5546) [reply msg/helpermsg/Heading.msg:1]: That wasn't a command, it was autocomplete (press tab)

Comment (patrick-5546) [reply msg/helpermsg/Heading.msg:1]: I think it would still be good to have some way to distinguish interfaces and their helpers if possible, here are my initial thoughts:

1. ~Do msgs have to be capitalized? What if we make interfaces uppercase and helpers lowercase?~ Not too sure about this, looking around I don't see anywhere where people have made lowercase messages. In fact, that may be confusing: https://answers.ros.org/question/328729/ros-msg-definitions-difference-between-std_msgstype-and-type/?answer=328730#post-id-328730

2. Prefix interfaces with `Inter`? And/or prefix helpers with `Helper`?

**Review by patrick-5546 — COMMENTED** [review]
Some more to dos:
- [x] Verify with @DFriend01 which msgs are internal to boat simulator and delete those (or create PR in boat_simulator to add those)
**Notes: Confirmed that ```Vector3``` and ```Kinematics``` are internal to boat simulator** - CL
- [x] Add comment to describe every field that isn't a custom message (e.g., `uint8`, `int32`)
- [x] Think of ways to distinguish interface and helper msgs and discuss with @patrick-5546 
**(Edit discussion 2023-06-16: Prefix ```Helper``` only, rest don't change) - CL**

- [x] Add to README the name, publishers, and subscribers of each interface/topic
- [x] After PR is merged, create PR in sailbot workspace to switch custom_interfaces branch/version to `main` in `src/new_project.repos`
- [x] Create issue for https://github.com/UBCSailbot/custom_interfaces/pull/2#issuecomment-1590200460

**Review by patrick-5546 — APPROVED** [review]
Great work on this PR

**Pull request merged**
